File: lib/domain/internal_jobs/gateways/scheduler_weather_farm_list_gateway.rb
# frozen_string_literal: true

module Domain
  module InternalJobs
    module Gateways
      module SchedulerWeatherFarmListGateway
        def list_reference_farms_for_weather_update
          raise NotImplementedError
        end

        def list_user_farms_for_weather_update
          raise NotImplementedError
        end
      end
    end
  end
end

File: crates/agrr-domain/src/internal_jobs/gateways/scheduler_weather_farm_list_gateway.rs
use crate::internal_jobs::dtos::SchedulerWeatherFarmRow;

/// Narrow read for scheduler weather update jobs (Rails `Update*WeatherDataJob` scopes).
pub trait SchedulerWeatherFarmListGateway: Send + Sync {
    /// `Farm.reference` with non-null latitude/longitude.
    fn list_reference_farms_for_weather_update(
        &self,
    ) -> Result<Vec<SchedulerWeatherFarmRow>, String>;

    /// `Farm.user_owned` with `weather_location_id` set.
    fn list_user_farms_for_weather_update(&self) -> Result<Vec<SchedulerWeatherFarmRow>, String>;

    /// User-owned farms with coordinates still awaiting initial full history fetch.
    fn list_user_farms_pending_initial_weather_fetch(&self) -> Result<Vec<i64>, String>;
}
